backdoor/src/main/java/org/codice/acdebugger/backdoor/Backdoor.java

```diff
@@ -0,0 +1,479 @@
+   * @param permission the permission to be granted
+   */
+  @SuppressWarnings({
+    "squid:S1181", /* letting VirtualMachineErrors bubble out directly, so ok to catch Throwable */
+    "squid:S1148" /* don't have access to logger at this stage */
+  })
+  public void grantPermission(String bundle, String permission) throws Exception {
+    try {
+      AccessController.doPrivileged(
+          (PrivilegedExceptionAction<Void>)
+              () -> {
+                synchronized (this) { // making sure only one thread at a time is granting
+                  final PermissionService permissionService = permServiceTracker.getService();
+
+                  if (permissionService != null) {
+                    permissionService.grantPermission(bundle, permission);
+                  }
+                  return null;
+                }
+              });
+    } catch (PrivilegedActionException e) {
+      e.getException().printStackTrace(); // suppress checkstyle:RegexpSingleline|RegexpMultiline
+      throw e.getException();
+    } catch (VirtualMachineError e) {
+      throw e;
+    } catch (Throwable t) {
+      t.printStackTrace(); // suppress checkstyle:RegexpSingleline|RegexpMultiline
+      throw t;
+    }
+  }
+
+  /**
+   * Checks if a domain has a given permission.
+   *
+   * <p>This method is called from the AC debugger and is purposely defined using Object to avoid
+   * having the classes not yet loaded when the AC debugger attempts to invoke the method.
+   *
+   * @param domain the domain to check for
+   * @param permission the permission to check for
+   * @return <code>true</code> if the domain has the specified permission; <code>false</code> if not
+   * @throws IllegalArgumentException if <code>domain</code> is not a {@link ProtectionDomain} or if
+   *     <code>serviceEvent</code> is not a {@link ServiceEvent}
+   */
+  @SuppressWarnings({
+    "squid:S1181", /* letting VirtualMachineErrors bubble out directly, so ok to catch Throwable */
+    "squid:S1148" /* don't have access to logger at this stage */
+  })
+  public boolean hasPermission(Object domain, Object permission) {
+    try {
+      if (!(domain instanceof ProtectionDomain)) {
+        throw new IllegalArgumentException("not a protection domain: " + domain);
+      }
+      if (!(permission instanceof Permission)) {
+        throw new IllegalArgumentException("not a permission: " + permission);
+      }
+      return AccessController.doPrivileged(
+          (PrivilegedAction<Boolean>)
+              () -> ((ProtectionDomain) domain).implies((Permission) permission));
+    } catch (VirtualMachineError e) {
+      throw e;
+    } catch (Throwable t) {
+      t.printStackTrace(); // suppress checkstyle:RegexpSingleline|RegexpMultiline
+      throw t;
+    }
+  }
+
+  /**
+   * Gets a bundle location for the given object. The object can be a bundle, a classloader, a
+   * protection domain, a bundle context, or even a classloader. This methods makes all attempts
```

Comment:
Classloader is mentioned twice here. Was it meant to be something else or just a typo?